src/components/rocp_sdk/rocp_sdk.c

```
@@ -488,19 +488,32 @@ hsa_is_enabled( void )
              (hsa_agent_get_infoPtr != NULL) );
 }
 
-static int
-load_hsa_sym( char *err_msg )
+void *dlopen_from_paths( char *libname, int pathCount, const char *paths[] )
 {
-    char pathname[PATH_MAX] = "libhsa-runtime64.so";
-    char *rocm_root = getenv("PAPI_ROCP_SDK_ROOT");
-    if (rocm_root != NULL) {
-        int count = snprintf(pathname, PATH_MAX, "%s/lib/libhsa-runtime64.so", rocm_root);
+    void *so = NULL;
+    for (int i = 0; i < pathCount; i++) {
+        if ( paths[i] == NULL ) {
+            continue;
+        }
+        char newpath[PATH_MAX];
+        int count = snprintf(newpath, PATH_MAX, "%s/lib/%s", paths[i], libname);
         if (count >= PATH_MAX) {
             SUBDBG("Status string truncated.");
         }
+        so = dlopen(newpath, RTLD_NOW | RTLD_GLOBAL);
+        if (so) {
+            return so;
+        }
```

Since you check for the path being empty, if the execution reached this point, then the path was wrong. We should error, instead of silently ignoring the bad path, using something like this:
 
else{
    return NULL;
}

---

That's true. I was going by Steve Kaufmann's request to fail when the user has provided a bad path, but this doesn't apply to the HSA library because we are not sure a priori where it will be. Long story short, your code is good the way it is.